### PR TITLE
Added check to verify cached file hash

### PR DESCRIPTION
**Pull request by SDkie**
Any interrupt in pull command will create an invalid cached file
Added check to verify cached file hash

Attn: @singularity-maintainers

### CODE REVIEW

File: internal/pkg/client/cache/library.go
Comment (ikaneshiro) [inline internal/pkg/client/cache/library.go]: This seems like an error that should be returned to the caller instead of just being part of debug output

File: internal/pkg/client/cache/library.go
Comment (ikaneshiro) [inline internal/pkg/client/cache/library.go]: I don't think we should be removing files from the cache in this function, just checking if something exists or not with the correct checksum. I think `cmd/internal/cli/pull.go` is a good place to remove/overwrite an invalid cache entry and it could be done simply by changing the force option of `client.DownloadImage()` to `true`

File: internal/pkg/client/cache/library.go
Comment (ikaneshiro) [inline internal/pkg/client/cache/library.go]: Probably shouldn't delete the image if we run into an error checking the hash

**Review by ikaneshiro — CHANGES_REQUESTED** [review]
Hey @SDkie this is looking good. Just some comments about error handling and modifying the cache in the `LibraryImageExists()` function

File: internal/pkg/client/cache/library.go
Comment (SDkie) [reply internal/pkg/client/cache/library.go]: @ikaneshiro : whether failing to get ImageHash will fail the pull command itself?

**Comment by SDkie** [comment]
@ikaneshiro: I made the changes you suggested. 
Also, the failure of the integration test is because of some network issue, I tested that locally.

File: internal/pkg/client/cache/library.go
Comment (ikaneshiro) [inline internal/pkg/client/cache/library.go]: I see how returning another bool to indicate if the file has the correct checksum, but I don't think its necessary for this function. I think we can just return `false` in the case that there is an image at the correct location and the hash doesn't match. This indicates that the image with the correct checksum doesn't exist in the cache regardless of if something with the correct filename exists in the cache. This will require the caller to know that there could be a file at that location in the cache even if this function returns `false`. This can be handled by having the force flag set when downloading an image into the cache, since it will either create a new file or overwrite an existing file with an incorrect checksum.

File: cmd/internal/cli/actions.go
Comment (ikaneshiro) [inline cmd/internal/cli/actions.go]: Here we can just hard code the `force` option to `true` since we always want to download an image in this location regardless of if its empty or there's an image there with an incorrect checksum. The same should apply to this function call in `pull.go` and `conveyorPacker_library.go`

**Review by ikaneshiro — CHANGES_REQUESTED** [review]
@SDkie One more tweak and I think it will be good to go! With respect to having another function that supports cache downloads explicitly, we're currently trying to restructure the code base so that we can have something like that. With the code base as is, doing that in the the `libexec` package would cause a cyclic import, so for the moment I think its best to leave it in this state while we move code around.

**Comment by SDkie** [comment]
@ikaneshiro : I have done the changes you suggested

File: cmd/internal/cli/actions.go
Comment (ikaneshiro) [inline cmd/internal/cli/actions.go]: Sorry, one last nitpick. Lets leave libraryURL hardcoded as `https://library.sylabs.io` since action commands don't currently support a custom library url. Another option is to change line 127 to use `PullLIbraryURI`, but I think the former is clearer to a reader

**Review by ikaneshiro — CHANGES_REQUESTED** [review]
Missed one small nitpick in my last review. Looks good to merge after this

**Review by ikaneshiro — APPROVED** [review]
LGTM. Will update with master once it passes CI and then merge

**Pull request merged**
